Mention how to enable tapping in the driver manpage

Index: xserver-xorg-input-synaptics/man/synaptics.man
===================================================================
--- xserver-xorg-input-synaptics.orig/man/synaptics.man	2009-05-13 00:14:36.753651106 +0900
+++ xserver-xorg-input-synaptics/man/synaptics.man	2009-05-13 00:27:28.785651827 +0900
@@ -813,6 +813,9 @@
 vertical scrolling is enabled, horizontal two-finger scrolling is disabled and
 edge scrolling is disabled. If no multi-finger capabilities are reported,
 edge scrolling is enabled for both horizontal and vertical scrolling.
+Tapping is disabled by default for touchpads with two buttons or more, to enable
+it you need to map tap actions to buttons. See the "TapButton1", "TapButton2"
+and "TapButton3" option descriptions for their meaning.
 .LP
 Button mapping for physical buttons is handled in the server.
 If the device is switched to left-handed (an in-server mapping of physical
