Index: xserver-xorg-input-synaptics/src/properties.c
===================================================================
--- xserver-xorg-input-synaptics.orig/src/properties.c	2008-09-29 10:07:16.000000000 +0300
+++ xserver-xorg-input-synaptics/src/properties.c	2008-09-29 10:12:48.000000000 +0300
@@ -28,7 +28,6 @@
 #endif
 
 #include "xf86Module.h"
-#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) >= 3
 
 #include <X11/Xatom.h>
 #include <xf86Xinput.h>
@@ -508,5 +507,4 @@
     return TRUE;
 }
 
-#endif
 
Index: xserver-xorg-input-synaptics/src/synaptics.c
===================================================================
--- xserver-xorg-input-synaptics.orig/src/synaptics.c	2008-09-29 10:12:39.000000000 +0300
+++ xserver-xorg-input-synaptics/src/synaptics.c	2008-09-29 10:13:30.000000000 +0300
@@ -125,10 +125,8 @@
 static Bool DeviceClose(DeviceIntPtr);
 static Bool QueryHardware(LocalDevicePtr);
 
-#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) >= 3
 void InitDeviceProperties(LocalDevicePtr local);
 int SetProperty(DeviceIntPtr dev, Atom property, XIPropertyValuePtr prop);
-#endif
 
 InputDriverRec SYNAPTICS = {
     1,
@@ -726,10 +724,8 @@
     if (!alloc_param_data(local))
 	return !Success;
 
-#if GET_ABI_MAJOR(ABI_XINPUT_VERSION) >= 3
     InitDeviceProperties(local);
     XIRegisterPropertyHandler(local->dev, SetProperty, NULL, NULL);
-#endif
 
     return Success;
 }
