Index: xserver-xorg-input-synaptics/fdi/11-x11-synaptics.fdi
===================================================================
--- xserver-xorg-input-synaptics.orig/fdi/11-x11-synaptics.fdi	2008-10-13 21:15:54.000000000 +1100
+++ xserver-xorg-input-synaptics/fdi/11-x11-synaptics.fdi	2008-10-13 21:16:00.000000000 +1100
@@ -18,7 +18,6 @@
       </match>
       <match key="info.product" contains="bcm5974">
         <merge key="input.x11_driver" type="string">synaptics</merge>
-        <merge key="input.x11_options.SHMConfig" type="string">true</merge>
         <merge key="input.x11_options.LeftEdge" type="string">0</merge>
         <merge key="input.x11_options.RightEdge" type="string">1280</merge>
         <merge key="input.x11_options.TopEdge" type="string">0</merge>
