Index: xserver-xorg-input-synaptics/src/eventcomm.c
===================================================================
--- xserver-xorg-input-synaptics.orig/src/eventcomm.c	2011-04-05 11:02:20.297458106 -0400
+++ xserver-xorg-input-synaptics/src/eventcomm.c	2011-04-05 11:02:48.753536302 -0400
@@ -547,6 +547,7 @@
     case EV_SYN:
         switch (ev->code) {
         case SYN_REPORT:
+            ProcessTouch(pInfo, priv);
             if (priv->has_touch && ecpriv->active_touches < 2)
                 hw->numFingers = ecpriv->active_touches;
             else if (comm->oneFinger)
@@ -558,7 +559,6 @@
             else
                 hw->numFingers = 0;
             *hwRet = *hw;
-            ProcessTouch(pInfo, priv);
             return TRUE;
         }
     case EV_KEY:
