--- a/src/synaptics.c
+++ b/src/synaptics.c
@@ -581,7 +581,7 @@ static void set_default_parameters(Input
     pars->press_motion_max_z = xf86SetIntOption(opts, "PressureMotionMaxZ", pressureMotionMaxZ);
     pars->resolution_detect  = xf86SetBoolOption(opts, "ResolutionDetect", TRUE);
 
-    pars->min_speed = xf86SetRealOption(opts, "MinSpeed", 0.4);
+    pars->min_speed = xf86SetRealOption(opts, "MinSpeed", 0.25);
     pars->max_speed = xf86SetRealOption(opts, "MaxSpeed", 0.7);
     pars->accl = xf86SetRealOption(opts, "AccelFactor", accelFactor);
     pars->trackstick_speed = xf86SetRealOption(opts, "TrackstickSpeed", 40);
