--- a/src/synaptics.c
+++ b/src/synaptics.c
@@ -1154,6 +1154,7 @@
     priv->lastButtons = 0;
     priv->prev_z = 0;
     priv->prevFingers = 0;
+    priv->num_active_touches = 0;
 #ifdef HAVE_MULTITOUCH
     memset(priv->open_slots, 0, priv->num_slots * sizeof(int));
 #endif
@@ -3119,6 +3120,7 @@
         if (hw->slot_state[i] == SLOTSTATE_OPEN) {
             priv->open_slots[priv->num_active_touches] = i;
             priv->num_active_touches++;
+            BUG_WARN(priv->num_active_touches > priv->num_slots);
         }
         else if (hw->slot_state[i] == SLOTSTATE_CLOSE) {
             Bool found = FALSE;
